File: Back/.Net/Modules/Users/BetFriends.Users.Infrastructure/Gateways/FakeAuthenticationGateway.cs
using BetFriends.Users.Application.Abstractions;
using BetFriends.Users.Application.Features.SignIn;
using BetFriends.Users.Infrastructure.Repositories;

namespace BetFriends.Users.Infrastructure.Gateways;

internal class FakeAuthenticationGateway(InMemoryUserRepository inMemoryUserRepository) : IAuthenticationGateway
{
    public Task<Authentication> AuthenticateAsync(AuthenticationRequest authenticationRequest)
    {
        var user = inMemoryUserRepository.GetUser(authenticationRequest.Email, authenticationRequest.Password);
        if (user is not null)
            return Task.FromResult(new Authentication("accesstoken", user.Snapshot.RefreshToken));

        return Task.FromResult<Authentication>(default!);
    }
}

File: Back/NodeJS/modules/users/src/infrastructure/FakeAuthenticationGateway.ts
import { IAuthenticationGateway } from "../application/abstractions/IAuthenticationGateway"
import { Authenticate } from "../application/features/sign-in/SignInHandler";
import { FakeUserRepository } from "./repositories/FakeUserRepository";
import { IJwtTokenGenerator } from "./IJwtTokenGenerator";

export class FakeAuthenticationGateway implements IAuthenticationGateway {
    constructor(private readonly userRepository: FakeUserRepository,
                private readonly jwtTokenGenerator: IJwtTokenGenerator
    ) {}
    Authenticate(email: string, password: string): Promise<Authenticate> {
        const user = this.userRepository.GetUsers().find(u => {
            const snapshot = u.GetSnapshot();
            return snapshot.Email == email && snapshot.Password == password;
        });
        if(user) {
            const userSnapshot = user.GetSnapshot();
            return Promise.resolve({
                AccessToken: this.jwtTokenGenerator.Generate({
                    Email: userSnapshot.Email,
                    Username: userSnapshot.Username,
                    UserId: userSnapshot.UserId
                }),
                RefreshToken: userSnapshot.RefreshToken
            });
        }
        return Promise.resolve(undefined!);
    }

}